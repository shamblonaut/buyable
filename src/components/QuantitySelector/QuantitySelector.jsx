import { useEffect, useState } from "react";
import { Minus, Plus } from "lucide-react";
import PropTypes from "prop-types";

import {
  Selector,
  DecrementButton,
  IncrementButton,
  CurrentQuantity,
} from "./QuantitySelector.styles";

const QuantitySelector = ({ quantity, setQuantity, decrement, increment }) => {
  const [value, setValue] = useState(quantity.toString());

  const resetValue = () => setValue(quantity.toString());

  const handleValueChange = (event) => {
    setValue(event.target.value);
    clearTimeout(resetValue);

    if (
      event.target.value !== "" &&
      !isNaN(event.target.value) &&
      Number(event.target.value) >= 0
    ) {
      setQuantity(Number(event.target.value));
    }
  };

  useEffect(resetValue, [quantity]);

  return (
    <Selector aria-label="cart-quantity-selector">
      <DecrementButton
        onClick={decrement}
        aria-label="quantity-decrement-button"
      >
        <Minus />
      </DecrementButton>
      <CurrentQuantity
        type="text"
        aria-label="custom-quantity-input"
        value={value}
        onChange={handleValueChange}
        onBlur={resetValue}
      />
      <IncrementButton
        onClick={increment}
        aria-label="quantity-increment-button"
      >
        <Plus />
      </IncrementButton>
    </Selector>
  );
};

QuantitySelector.propTypes = {
  quantity: PropTypes.number.isRequired,
  setQuantity: PropTypes.func.isRequired,
  decrement: PropTypes.func.isRequired,
  increment: PropTypes.func.isRequired,
};

export default QuantitySelector;
