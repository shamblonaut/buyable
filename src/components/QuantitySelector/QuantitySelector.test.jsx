import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import QuantitySelector from "./QuantitySelector";

const mockQuantity = 4;
const mockSetQuantity = vi.fn();
const mockDecrement = vi.fn();
const mockIncrement = vi.fn();

describe("QuantitySelector component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly when cart is empty", () => {
    const { container } = render(
      <QuantitySelector
        quantity={mockQuantity}
        setQuantity={mockSetQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("calls allocated decrement function", async () => {
    const user = userEvent.setup();

    let decrement = vi.spyOn(
      {
        decrement: vi.fn(),
      },
      "decrement",
    );
    let increment = vi.spyOn(
      {
        increment: vi.fn(),
      },
      "increment",
    );

    render(
      <QuantitySelector
        quantity={mockQuantity}
        setQuantity={mockSetQuantity}
        decrement={decrement}
        increment={increment}
      />,
    );

    await user.click(
      screen.getByRole("button", { name: "quantity-decrement-button" }),
    );

    expect(decrement).toHaveBeenCalled();
    expect(increment).not.toHaveBeenCalled();
  });

  it("calls allocated increment function", async () => {
    const user = userEvent.setup();

    let decrement = vi.spyOn(
      {
        decrement: vi.fn(),
      },
      "decrement",
    );
    let increment = vi.spyOn(
      {
        increment: vi.fn(),
      },
      "increment",
    );

    render(
      <QuantitySelector
        quantity={mockQuantity}
        setQuantity={mockSetQuantity}
        decrement={decrement}
        increment={increment}
      />,
    );

    await user.click(
      screen.getByRole("button", { name: "quantity-increment-button" }),
    );

    expect(increment).toHaveBeenCalled();
    expect(decrement).not.toHaveBeenCalled();
  });

  it("passes (only) custom number quantity input to setQuantity", async () => {
    const user = userEvent.setup();

    let setQuantity = vi.spyOn(
      {
        setQuantity: vi.fn(),
      },
      "setQuantity",
    );

    render(
      <QuantitySelector
        quantity={mockQuantity}
        setQuantity={setQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );

    expect(setQuantity).not.toHaveBeenCalled();

    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "16",
    );

    expect(setQuantity).toHaveBeenCalledWith(16);
    vi.clearAllMocks();

    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "asdf",
    );

    expect(setQuantity).not.toHaveBeenCalled();
  });

  it("resets the custom input value when a new quantity value is passed", async () => {
    const user = userEvent.setup();

    const { rerender } = render(
      <QuantitySelector
        quantity={mockQuantity}
        setQuantity={mockSetQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "16",
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("16");

    rerender(
      <QuantitySelector
        quantity={7}
        setQuantity={mockSetQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("7");
  });

  it("resets the custom input value when focus is lost on an invalid input", async () => {
    const user = userEvent.setup();

    let quantity = mockQuantity;
    let setQuantity = vi.spyOn(
      {
        setQuantity: (newQuantity) => {
          quantity = newQuantity;
        },
      },
      "setQuantity",
    );

    const { rerender } = render(
      <QuantitySelector
        quantity={quantity}
        setQuantity={setQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    // Empty value
    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("");

    await user.tab();
    rerender(
      <QuantitySelector
        quantity={quantity}
        setQuantity={setQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue(mockQuantity.toString());

    // Alphabetic characters
    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "asdf",
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("asdf");

    await user.tab();
    rerender(
      <QuantitySelector
        quantity={quantity}
        setQuantity={setQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue(mockQuantity.toString());

    // Negative number
    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "-9",
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("-9");

    await user.tab();
    rerender(
      <QuantitySelector
        quantity={quantity}
        setQuantity={setQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue(mockQuantity.toString());

    // Valid number
    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "7",
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("7");

    await user.tab();
    rerender(
      <QuantitySelector
        quantity={quantity}
        setQuantity={setQuantity}
        decrement={mockDecrement}
        increment={mockIncrement}
      />,
    );

    expect(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    ).toHaveValue("7");
    expect(setQuantity).toHaveBeenCalledWith(7);
  });
});
