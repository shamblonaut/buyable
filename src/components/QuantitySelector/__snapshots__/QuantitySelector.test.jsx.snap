// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuantitySelector component > renders correctly when cart is empty 1`] = `
<div>
  <div
    aria-label="cart-quantity-selector"
    class="sc-bRKDuR eFzYWC"
  >
    <button
      aria-label="quantity-decrement-button"
      class="sc-hvigdm sc-fhHczv cSvOhk eYhuUE"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-minus"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
      </svg>
    </button>
    <input
      aria-label="custom-quantity-input"
      class="sc-dTvVRJ hypWRI"
      type="text"
      value="4"
    />
    <button
      aria-label="quantity-increment-button"
      class="sc-hvigdm sc-ggWZvA cSvOhk bgUnbA"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
    </button>
  </div>
</div>
`;
